---
title: "Results of trait-assisted GS"
author: "Quentin D. Read"
date: "`r format(Sys.time(), '%d %B %Y')`"
output: 
  html_document:
      css: "gtstyle.css"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE, echo = FALSE)
```

## Change log

- 07 March: add rank correlations and analysis of rank-transformed Brix scores
- 03 March: first version

```{r read data}
library(tidyverse)
library(gt)
library(readxl)
library(data.table)

metrics_tags <- read_csv('project/output/TAGS_all_metrics.csv')

physical_traits <- c("stkwt_kg", "diam", "Brix", "Fiber", "Pol", "Sucrose", "Purity", "stalk_ha")
economic_traits <- c("TCH", "TRS", "CRS", "TSH", "EI")

theme_set(theme_bw() +
            theme(strip.background = element_blank(),
                  legend.position = 'bottom'))

fill_scale <- scale_fill_brewer(palette = 'Dark2')
null_line <- geom_hline(yintercept = 0.167, size = 1.2, linetype = 'dashed')
null_line2 <- geom_hline(yintercept = 0, size = 1.2, linetype = 'dashed')

# Read each sheet from phenotype XLSX and combine to single data frame
# Note there are multiple different NA flags in the data
pheno_file <- 'project/data/Phenotype_updated_2017-18_IL_analysis_120921.xlsx'
sheet_names <- excel_sheets(pheno_file)
phenotype_sheets <- map(sheet_names, ~ cbind(crop_cycle = ., read_xlsx(pheno_file, sheet = ., na = c('.', '-', '-!'))))

# Correct column name, changing Diam to diam (bugfix)
for (i in 1:length(phenotype_sheets)) names(phenotype_sheets[[i]])[names(phenotype_sheets[[i]]) == 'Diam'] <- 'diam'

phenotypes <- rbindlist(phenotype_sheets, fill = TRUE)

phenotypes_long <- phenotypes[!crop_cycle %in% 'Ratoonability', mget(c("crop_cycle", "Clone", "Rep", "Row", "Column", physical_traits, economic_traits))] |>
  melt(id.vars = c("crop_cycle", "Clone", "Rep", "Row", "Column"), variable.name = 'trait')
```

## Summary

I did trait-assisted genomic selection using the `sommer` package in R. This is a mixed model with a multivariate response that uses the additive genetic relationship matrix as part of the random effects. Each genotype has multiple response variables which are a single trait measured in >1 crop cycle. I did this for the first ratoon, where plant cane and first ratoon trait BLUPs were used as the response variable, and for the second ratoon, where plant cane, first ratoon, and second ratoon were used. I excluded the test-set BLUPs for the holdout set in the cross-validation fold *only* for the crop cycle to be predicted; the BLUPs for the "older" data were used for model fitting for both the training and test set. This simulates the situation where we have trait data from one crop cycle but don't yet have it for the next year's cycle.

I did 5-fold cross-validation for each of the two trait-assisted GS, and repeated the CV procedure 25 times.

In the following plots we see that the performance for the trait-assisted GS is not very good. In fact it is slightly worse than without including the additional trait information. I believe this may be because there isn't a very close relationship among traits across crop cycles within a genotype. Because of that, adding the other traits from older crop cycles gives no additional reliable signal and just some more noise. I would not recommend pursuing this much further for this particular population because the initial foray into trait-assisted GS has yielded such unpromising results.

## Prediction performance: physical traits

I only plotted the $r$ (observed-predicted correlation from the cross-validated data) as an index of performance here because it tells well enough that trait-assisted GS, at least as implemented in the `sommer` R package, does not do a very good job.

```{r}
ggplot(metrics_tags %>% filter(trait %in% physical_traits), aes(x = trait, y = r, fill = crop_cycle, group = interaction(trait, crop_cycle))) +
  geom_boxplot() +
  fill_scale +
  null_line2
```

## Prediction performance: economic traits

```{r}
ggplot(metrics_tags %>% filter(trait %in% economic_traits), aes(x = trait, y = r, fill = crop_cycle, group = interaction(trait, crop_cycle))) +
  geom_boxplot() +
  fill_scale +
  null_line2
```

## Look at correlations among crop cycles

My initial suspicion was that the trait-assisted GS had such poor results because of low correlation between crop cycles. I calculated the Pearson correlations to take a look. They are quite modest. So I am not surprised that it probably just added more noise and not enough reliable signal, leading to the worse results of trait-assisted GS relative to models without taking that covariation into account. But we do see that the traits where correlation between crop cycle is highest do tend to have slightly better performance of the trait-assisted GS (such as stalks per hectare and diameter).

```{r}
# Reshape so we have crop cycles in separate columns
phenotypes_cc <- dcast(phenotypes_long, Clone + Rep + Row + Column + trait ~ crop_cycle)

trait_corrs <- phenotypes_cc[, .(
    cor12 = cor(`Plant Cane_2017`, `Ratoon 1_2018`, use = 'pairwise.complete.obs'),
    cor13 = cor(`Plant Cane_2017`, `Ratoon 2_2019`, use = 'pairwise.complete.obs'),
    cor23 = cor(`Ratoon 1_2018`, `Ratoon 2_2019`, use = 'pairwise.complete.obs')
    ), by = .(trait)]

gt(trait_corrs) %>%
  tab_header(title = 'Trait correlations between crop cycles') %>%
  cols_label(cor12 = 'plant cane vs. 1st ratoon', cor13 = 'plant cane vs. 2nd ratoon', cor23 = '1st ratoon vs. 2nd ratoon') %>%
  fmt_number(columns = 2:4, decimals = 3) %>%
  data_color(columns = 2:4, colors = scales::col_bin(palette = 'Reds', domain = NULL))
```

## Rank correlations among crop cycles

Based on the discussion we had last week, I also calculated the rank (Spearman) correlations among crop cycles. The rank correlations are almost identical to the Pearson correlations so it's unlikely that there will be much difference if we fit trait-assisted GS to rank-transformed data but I will try it anyway, just for the example of Brix score, below.

```{r}
trait_rankcorrs <- phenotypes_cc[, .(
    cor12 = cor(`Plant Cane_2017`, `Ratoon 1_2018`, use = 'pairwise.complete.obs', method = 'spearman'),
    cor13 = cor(`Plant Cane_2017`, `Ratoon 2_2019`, use = 'pairwise.complete.obs', method = 'spearman'),
    cor23 = cor(`Ratoon 1_2018`, `Ratoon 2_2019`, use = 'pairwise.complete.obs', method = 'spearman')
    ), by = .(trait)]

gt(trait_rankcorrs) %>%
  tab_header(title = 'Trait rank correlations between crop cycles') %>%
  cols_label(cor12 = 'plant cane vs. 1st ratoon', cor13 = 'plant cane vs. 2nd ratoon', cor23 = '1st ratoon vs. 2nd ratoon') %>%
  fmt_number(columns = 2:4, decimals = 3) %>%
  data_color(columns = 2:4, colors = scales::col_bin(palette = 'Blues', domain = NULL))
```

## Trait-assisted GS on rank-transformed Brix

Set up everything:

```{r, echo = TRUE}
library(lme4)
library(furrr)
library(rrBLUP) 
library(sommer)
library(glue)

source('GS_sugarcane_2022_fns.R') # Load needed functions

genotypes <- fread('project/data/sugarcane.10501.SNPs.432.Inds.CloneNames.hmp.txt', sep = '\t')

# Find columns representing clones
id_columns <- c("rs.", "alleles", "chrom", "pos", "strand", "assembly.", "center", "protLSID", "assayLSID", "panelLSID", "QCcode")
clone_columns <- setdiff(names(genotypes), id_columns)

# Convert SNPs to numerical format ----------------------------------------

# Exclude rows with >2 alleles
genotypes <- genotypes[nchar(alleles) <= 3]
geno_mat_raw <- as.matrix(genotypes[, ..clone_columns])

# Get ID of first and second allele. Map them to 0 and 2 respectively. Map all other values (heterozygotes) within the row to 1.
alleles_list <- strsplit(genotypes$alleles, '/')
allele1 <- map_chr(alleles_list, 1)
allele2 <- map_chr(alleles_list, 2)

geno_mat <- map2(asplit(geno_mat_raw, 1), alleles_list, function(x, allele) fcase(x == allele[1], 0, x == allele[2], 2, default = 1))
geno_mat <- do.call(cbind, geno_mat) # Transpose the matrix so rows are individuals and columns SNPs

# Set row names of genotype matrix to names of clones
dimnames(geno_mat) <- list(clone_columns, NULL)
```

Get the BLUPs for rank-transformed Brix.

```{r, echo = TRUE}
pheno_brix <- phenotypes_long[trait == 'Brix']
pheno_brix[, value := rank(value), by = .(crop_cycle)]

brix_blups <- pheno_brix[, blup_trait(.SD, crop_cycle_fixef = FALSE), by = .(trait, crop_cycle)]
setnames(brix_blups, c('trait', 'crop_cycle', 'Clone', 'BLUP'))

# Remove the clone checks 
check_IDs <- c('CP96-1252', 'CP00-1101')
brix_blups <- brix_blups[!Clone %in% check_IDs]

# Change crop cycle names to the short name
brix_blups[, crop_cycle := fcase(
  crop_cycle == 'Plant Cane_2017', 'PlantCane',
  crop_cycle == 'Ratoon 1_2018', 'Ratoon1',
  crop_cycle == 'Ratoon 2_2019', 'Ratoon2'
)]

```

Do 5 iterations of trait-assisted GS with 5-fold CV for Brix.

```{r, echo = TRUE}
n_iter <- 5
n_folds <- 5

combos <- CJ(iter = 1:n_iter, trait = c('Brix'), crop_cycle = c('Ratoon1', 'Ratoon2'))

# Do the GS. Write observed and predicted phenotypes and prediction accuracy metrics with each iteration.
# If the prediction metric file already exists, skip that iteration (this allows the script to be rerun)
traitgs_pred_metrics <- future_pmap(combos, function(iter, trait, crop_cycle) {
  metric_file_name <- glue('project/output/BrixTAGS/metrics_{trait}_{crop_cycle}_{iter}.csv')
  if (!file.exists(metric_file_name)) {
    pred_vals <- trait_assisted_gs(
      GD = geno_mat, PD = brix_blups, crop_cycle_to_use = crop_cycle, trait_to_use = trait, k = n_folds
    )
    pred_metrics <- with(pred_vals, calc_metrics(Y_obs, Y_pred))
    fwrite(pred_metrics, metric_file_name)
  } else {
    pred_metrics <- fread(metric_file_name)
  }
  return(pred_metrics)
}, .options = furrr_options(seed = 222))

combos[, metrics := traitgs_pred_metrics]
metrics_brix_rank <- unnest_dt(combos, col = metrics, id = .(iter, trait, crop_cycle))
```

Look at the results. Compare the prediction performance for the rank-transformed Brix scores to the original performance using untransformed Brix scores. (Note I only did 5 iterations for the rank-transformed scores, just to get an idea of it.) The rank-transformation makes no difference; the spread of values is practically identical.

```{r}
setDT(metrics_tags)
metrics_brix_orig <- metrics_tags[trait == 'Brix']

metrics_brix_orig[, transformation := 'none']
metrics_brix_rank[, transformation := 'rank']

r_brix <- rbind(metrics_brix_orig, metrics_brix_rank)

ggplot(r_brix, aes(x = crop_cycle, y = r, group = interaction(crop_cycle, transformation), color = transformation)) +
  geom_point(alpha = 0.7, size = 2, position = position_dodge(width = 0.2))
```

